pkg/devfile/adapters/docker/component/adapter.go

```diff
@@ -0,0 +1,219 @@
+	return utils.ComponentExists(a.Client, cmpName)
+}
+
+func (a Adapter) createComponent() (err error) {
+	componentName := a.ComponentName
+
+	volumeLabels := map[string]string{
+		"component": componentName,
+		"type":      "projects",
+	}
+
+	supportedComponents := adaptersCommon.GetSupportedComponents(a.Devfile.Data)
+	if len(supportedComponents) == 0 {
+		return fmt.Errorf("No valid components found in the devfile")
+	}
+
+	// Create a docker volume to store the project source code
+	volume, err := a.Client.CreateVolume(volumeLabels)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to create project source volume for component %s", componentName)
+	}
+
+	projectVolumeName := volume.Name
+
+	// Loop over each component and start a container for it
+	for _, comp := range supportedComponents {
+		s := log.Spinner("Pulling image " + *comp.Image)
+		defer s.End(false)
+
+		// Pull the image (as the Docker daemon requires it to be on the system before starting it)
+		err = a.Client.PullImage(*comp.Image)
+		if err != nil {
+			return errors.Wrapf(err, "Unable to pull %s image", *comp.Image)
+		}
+		s.End(true)
+
+		// Start the container
+		err = a.startContainer(componentName, projectVolumeName, comp)
+		if err != nil {
+			return errors.Wrapf(err, "Unable to start container for devfile component %s", *comp.Alias)
+		}
+		glog.V(3).Infof("Successfully created container %s for component %s", *comp.Image, componentName)
+	}
+	glog.V(3).Infof("Successfully created all containers for component %s", componentName)
+
+	return nil
+}
+
+func (a Adapter) updateComponent() (err error) {
+	glog.V(3).Info("The component already exists, attempting to update it")
+	componentName := a.ComponentName
+
+	// Get the project source volume
+	volumeLabels := map[string]string{
+		"component": componentName,
+		"type":      "projects",
+	}
+	vols, err := a.Client.GetVolumesByLabel(volumeLabels)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to retrieve source volume for component "+componentName)
+	}
+	if len(vols) == 0 {
+		return fmt.Errorf("Unable to find source volume for component %s", componentName)
+	}
+	projectVolumeName := vols[0].Name
+
+	supportedComponents := adaptersCommon.GetSupportedComponents(a.Devfile.Data)
+	if len(supportedComponents) == 0 {
+		return fmt.Errorf("No valid components found in the devfile")
+	}
+
+	for _, comp := range supportedComponents {
+		// Check to see if this component is already running and if so, update it
+		// If component isn't running, re-create it, as it either may be new, or crashed.
+		containers, err := a.Client.GetContainersByComponentAndAlias(componentName, *comp.Alias)
+		if err != nil {
+			return errors.Wrapf(err, "unable to list containers for component %s", componentName)
+		}
+		if len(containers) == 0 {
+			// Container doesn't exist, so need to pull its image (to be safe) and start a new container
+			s := log.Spinner("Pulling image " + *comp.Image)
+			defer s.End(false)
```

Comment:
I understand this part that if this loop returns before reaching https://github.com/openshift/odo/pull/2806/files#diff-00b1316479a0f694654a918c0915c476R139. The spinner will end in failure but this is a bit confusing, so can we please move it inside the err != nil if statement?

Reply:
Sure, done.